panel/src/components/Dialogs/Dialog.vue

```diff
@@ -219,10 +219,7 @@ export default {
     focus() {
       if (this.$refs.dialog?.querySelector) {
         const btn = this.$refs.dialog.querySelector(".k-dialog-button-cancel");
-
-        if (typeof btn?.focus === "function") {
-          btn.focus();
-        }
+        btn?.focus?.();
```

Comment:
I'm not familiar with javascript but what if `btn.focus` is not a `function`? 🤔

Reply:
That's why TypeScript is recommended for Kirby 4. 😜
The query selector will return a `HTMLElement`, which has the focus method built-in: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus

Reply:
OK for `querySelector`. What about other optional chaining chages where you made for function types?

Reply:
If it's a `ref`, then it's a `HTMLElement` as well.

Reply:
On the other elements where `selector[method]` applies, only `select`, `blur` and `focus` are the values used. Thus, we can skip the function check. If this method exists on the element, then it is intrinsically a function. Correct me if I'm wrong, of course. 🙂

Reply:
Personally, I find `focus?.()` a bit weird. I know we need it in some places (e.g. I think in the overlay it's not clear what element we are calling this on), but in most of these places we should know what element we are referring/referencing and we should know if it has a focus method or not.

Reply:
Weird it may be, but safer it is. 😉 TypeScript will help you from Kirby 4 onwards. 😇

Reply:
As far as I can see, Typescript isn't really an option with Vue 2

Reply:
Kirby 4 would be based on Vue 3, I guess... 👀

Reply:
Yea, just thinking out loud here: I don't think we will ever be really able to do the big lift as from v2 to v3, rewriting everything for more than a year. So I'm still looking at ways to rather gradually introduce and move from Vue 2 to Vue 3 or from JS to TypeScript. But every time it feels more like they want us to pull the rug out beneath our feet.

Reply:
Thanks for sharing your thoughts. Maybe rewriting using TS and Composition API won't be important and is optional. But one day, the jump to Vue 3 (without rewriting the entire panel, "only" the necessary parts) will probably make sense. Simpler plugins may just have to be recompiled with kirbyup (would then introduce a v3 option). More complex ones would probably still be tricky to migrate.